Added Tutorial writeup - Segment Tree

Dhruvil-Lakhtaria opened the pull request:
## Change Summary

Issue Number - #215 
Added Tutorial and Solution for Segment Tree and Lazy Propagation. The Dynamic Segment tree part is not that much extensive and has less resources and question. I hope the above changes are enough to be added and fine. Can you please review that much. :).

## Checklist

**Ideally all of them should be checked. If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed.**

### General 

- [X] Make sure you've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md) 
- [X] Make sure you've followed the template format
- [X] Make sure you've started the app locally and verified all the content and all links (if applicable) must be accessible correctly
- [X] Make sure you've applied LaTex for variables, formulas and time / space complexity instead of using backticks
- [X] Make sure you've formatted your code well (Code Block is rendered using `<Tabs/>`)
      Below is a C++ example. You can include multiple langs using `<TabItem/>` if necessary
      
````
<Tabs>

<TabItem value="cpp" label="C++">
<SolutionAuthor name="@YOUR_ALIAS"/>

```cpp
// Your code goes here
// Prefer K&R Coding style
// See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example
```

</TabItem>

</Tabs>
````
- [X] Make sure you've explained your code well (write comments above each line - not inline)
- [X] Make sure you've written your explanation well and it is easy to understand for beginners
- [X] Make sure you've included Time & Space Complexity Analysis
- [X] Make sure you've included a meaningful approach name for your solution. Don't leave it blank. e.g. `## Approach 1: Two Pointers`.

### For Tutorial Articles

- [X] Make sure you've explained your topic using 1 - 3 LC problems rather than using external problems. Solutions must be taken with explanations as well.
- [X] Make sure you've provided suggested problems at the end using the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [X] Make sure you've given credits / references if you use external resources. For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).

wingkwong reviewed (changes requested):
1. suggest to combine the tutorials and put them under `tutorials/advanced-topics`.
2. see https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for coding style
3. comments should start with lower case and after a space
4. wrong format for code block. use `<Tabs>`. See the checklist.
5. not enough explanations. beginners won't be able to follow

Dhruvil-Lakhtaria commented:
> 1. suggest to combine the tutorials and put them under `tutorials/advanced-topics`.
> 2. see https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for coding style
> 3. comments should start with lower case and after a space
> 4. wrong format for code block. use `<Tabs>`. See the checklist.
> 5. not enough explanations. beginners won't be able to follow

Dhruvil-Lakhtaria commented:
Any updates.

wingkwong commented:
Is it ready for review? I didn't you re-request review.

wingkwong commented:
Oh I missed your comment.

> can you specify what exactly needs more explanation

Basically you just show the implementation with some comments. However, you didn't mention why we need segment tree and the details of how to build it (your comments in code are just explaining what that line does only). Besides, you should include the full solution to your example. 

By the way, just take a quick look, the code is a bit messy. Please format it well following [this guideline](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)

Dhruvil-Lakhtaria commented:
Added explaination for segement tree.(why segment tree and its functions).
Changed the question for lazy propagation.
Checked code with k&r style.
Solution added are full solutions.
Can you review again.

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+ 
+### Representation
+We can implement the above using an array representation.
+Each Index denotes the answer to that interval with $$(2*i+1) th$$ index as its left child and $$(2*i+2)th$$ index as its right child.
```
1. add space for each operator please. i.e. `$2*i + 1$` -> `$2 * i + 1$`

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+We can implement the above using an array representation.
+Each Index denotes the answer to that interval with $$(2*i+1) th$$ index as its left child and $$(2*i+2)th$$ index as its right child.
+
+![Segment-Tree](https://miro.medium.com/max/1100/1*9V1l8_weUfmT4jHg21EE5w.png)
```
please update the diagram. it's hard to read with blue background and black text color

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+The question asks us to do 2 operations with the query operation having the logic to get sum of that operation. Seeing this only we can decide to use a segment tree data structure to solve the problem.
+
+### Building Segment tree for this question. 
+We start from the root denoting the entire array. Now we need information of $(0 - mid)$ and $(mid + 1 - n)$ for building current root. so we recursively call them. We know left child id equivalent to $2*i + 1$ array index and right child is $2*i + 2$.
```
1. why 0 - mid and mid + 1 - n? 
2. add space for each operator please. i.e. `$2*i + 1$` -> `$2 * i + 1$`

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+The logic for point update is pretty much the same involving recursive calls and backtracking. We reach till a leaf node where we need to update the value and while backtracking all the nodes that come into path needs to be checked for updation.
+
+For example, Analyse the image below.
+![Point Update](https://cdn.codespeedy.com/wp-content/uploads/2019/12/Node-or-Point-updates.png)
```
1. don't use external diagram. if it is necessary, give credit
2. please use a diagram that matches your example test case.

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+Here we are update 2nd index and adding 3 to it.So all the nodes on the path back to root from it must also be updated.
+
+<Tabs>
+<TabItem value="CPP" label="CPP">
```
change to following (apply to all)

```
<TabItem value="cpp" label="c++">
```

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
````diff
@@ -0,0 +1,179 @@
+<TabItem value="CPP" label="CPP">
+<SolutionAuthor name="@DhruvilLakhtaria"/>
+
+```CPP
````
format your code well following [this guideline](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+</TabItem>
+</Tabs>
+
+
```
Missing full solution to your example.

Dhruvil-Lakhtaria replied:
The solution is complete only. Leetcode question is to only fill the functions, I have added a sample usage just incase someone wants to know how to use this segment tree class.

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+</Tabs>
+
+
+### Time Complexity and Space Complexity Analysis 
```
change to `Complexity Analysis`

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+>Time Complexity - $$O(4*N)$$ = $$O(N)$$;
+
+2. `update(i,val)`: Here we go recursively to the segment holding ith element and update it. On the way back (backtracking) we update all the node accordingly.
+> Time Complexity = $$O(logN)$$ as height of the tree is logN.
```
LaTex on logN

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+> Time Complexity = $$O(logN)$$ as height of the tree is logN.
+
+3. `query()`: We are given an range [L-R]. We basically find sub-segments that add up to required segment. We need to travel across the height atmax once to get our answer.
+> Time Complexity = $$O(logN)$$. as height of the tree is logN.
```
LaTex on logN

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+> 1. Get sum of values in range $$[L-R]$$.
+> 2. Update ith element to new value. 
+
+### Approach Using Segment Tree.
```
1. Missing Query part.
2. for each section, add a code snippet to explain. at the end, show the full segment tree template (which you did below already)

Dhruvil-Lakhtaria replied:
I dont think adding code snippet again here would help as the function name are clear enough and people can reaf there  itself.

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+
+    vector<int> v;
+    SegmentTree (vector<int>& arr, int n) {
+	    v = vector<int>(4*n);
```
As a beginner, they may ask why you need to reserve $4 * n$. Explain it well. Better with some proof.

wingkwong commented on tutorials/advanced-topics/basic-segment-tree.md:
```diff
@@ -0,0 +1,179 @@
+
+
+
+### Segment Tree Implementation for Range Min Query.
```
1. Your example is range sum, not range min.
2. Give some guideline how to change the template in order to find min / max / gcd instead of sum.

Dhruvil-Lakhtaria commented:
Made requested change and commented on few change requests.
Please review.

wingkwong commented:
Just have a quick look. It looks good. However, you haven't provided the full solution to the example [307. Range Sum Query - Mutable](https://leetcode.com/problems/range-sum-query-mutable/). The full solution means - I can copy it and run it on LC directly. Please add it back :)

I'll start reviewing lazy propagation one in the coming days. At the same time, feel free to supplement more as detailed as possible.

Dhruvil-Lakhtaria commented:
> Just have a quick look. It looks good. However, you haven't provided the full solution to the example [307. Range Sum Query - Mutable](https://leetcode.com/problems/range-sum-query-mutable/). The full solution means - I can copy it and run it on LC directly. Please add it back :)
> 
> I'll start reviewing lazy propagation one in the coming days. At the same time, feel free to supplement more as detailed as possible.

Added leetcode submitable solution.

wingkwong commented:
@Dhruvil-Lakhtaria Please format your code well using the guideline I gave. The indentation is not aligned as well.

Dhruvil-Lakhtaria commented:
made some changes

Dhruvil-Lakhtaria commented:
Can you look please merge the basic segment tree part after reviewing

wingkwong commented:
![image](https://user-images.githubusercontent.com/35857179/198646964-7074cb37-718d-4982-ba3d-7116dc642334.png)

I don't see the code is well formatted.

wingkwong commented on tutorials/advanced-topics/lazy-propagation.md:
```diff
@@ -0,0 +1,189 @@
+
+### Idea behind Lazy Propagation
+
+The idea behind the algorithm is don’t not update a node until needed which will avoid the repeated sharing.
+
+Updating the interval represented by a node is similar to update the subtree of that node in segment tree. Suppose we have to update subtree of **Node X**, instead of updating all the values at that instant we will only update that node and mark its children that their subtree has to be updated and for maintaining this record we will use **Lazy array**. Size of Lazy array would be same as the siz ...
```
https://medium.com/nybles/understanding-range-queries-and-updates-segment-tree-lazy-propagation-and-mos-algorithm-d2cd2f6586d8

Another copycat

Dhruvil-Lakhtaria replied:
I referred to this blog for learning and kept the same content for my blog.

wingkwong replied:
https://github.com/wingkwong/leetcode-the-hard-way/blob/main/.github/pull_request_template.md#checklist

>  This Pull Request is all my own work. (You'll be blacklisted if you are caught for plagiarism.)

Dhruvil-Lakhtaria replied:
I just copied the theory part rest solution and all are mine only

Dhruvil-Lakhtaria commented:
> ![image](https://user-images.githubusercontent.com/35857179/198646964-7074cb37-718d-4982-ba3d-7116dc642334.png)
> 
> I don't see the code is well formatted.

Is this regarding comments not formatted... I will change that
